cpp/src/io/statistics/statistics_type_identification.cuh

```diff
@@ -0,0 +1,244 @@
+
+ public:
+  // Does type T have an extrema?
+  static constexpr bool is_supported = std::is_arithmetic_v<T> or std::is_same_v<T, string_view> or
+                                       cudf::is_duration<T>() or cudf::is_timestamp<T>() or
+                                       cudf::is_fixed_point<T>();
+
+  using type = typename std::
+    conditional_t<std::is_arithmetic_v<T>, arithmetic_extrema_type, non_arithmetic_extrema_type>;
+
+  __device__ static type convert(const T& val)
+  {
+    if constexpr (std::is_arithmetic_v<T> or std::is_same_v<T, string_view>) {
+      return val;
+    } else if constexpr (cudf::is_fixed_point<T>()) {
+      return val.value();
+    } else if constexpr (cudf::is_duration<T>()) {
+      return val.count();
+    } else if constexpr (cudf::is_timestamp<T>()) {
+      return val.time_since_epoch().count();
+    } else {
+      static_assert(dependent_false<T>::value, "aggregation_type does not exist");
+    }
+    return type{};
+  }
+};
+
+template <typename T>
+class aggregation_type {
+ private:
+  using integral_aggregation_type =
+    typename std::conditional_t<std::is_signed_v<T>, int64_t, uint64_t>;
+
+  using arithmetic_aggregation_type =
+    typename std::conditional_t<std::is_integral_v<T>, integral_aggregation_type, double>;
+
+  using non_arithmetic_aggregation_type =
+    typename std::conditional_t<cudf::is_fixed_point<T>() or cudf::is_duration<T>() or
+                                  cudf::is_timestamp<T>()  // To be disabled with static_assert
+                                  or std::is_same_v<T, string_view>,
+                                int64_t,
+                                void>;
+
+  // unsigned int/bool -> uint64_t
+  // signed int        -> int64_t
+  // float/double      -> double
+  // decimal32/64      -> int64_t
+  // duration_[T]      -> int64_t
+  // string_view       -> int64_t
+  // NOTE : timestamps do not have an aggregation type
+
+ public:
+  // Does type T aggregate?
+  static constexpr bool is_supported = std::is_arithmetic_v<T> or std::is_same_v<T, string_view> or
+                                       cudf::is_duration<T>() or cudf::is_fixed_point<T>();
+
+  using type = typename std::conditional_t<std::is_arithmetic_v<T>,
+                                           arithmetic_aggregation_type,
+                                           non_arithmetic_aggregation_type>;
+
+  __device__ static type convert(const T& val)
+  {
+    if constexpr (std::is_same_v<T, string_view>) {
+      return val.size_bytes();
+    } else if constexpr (std::is_integral_v<T>) {
+      return val;
+    } else if constexpr (std::is_floating_point_v<T>) {
+      return isnan(val) ? 0 : val;
+    } else if constexpr (cudf::is_fixed_point<T>()) {
+      return val.value();
+    } else if constexpr (cudf::is_duration<T>()) {
+      return val.count();
+    } else {
+      static_assert(dependent_false<T>::value, "aggregation_type for timestamps do not exist");
```

Comment:
My intention was to fail compilation if aggregation_type is used with timestamps since they do not aggregate. Should the warning message not specify the timestamp type?

Reply:
My concern is that the error message does not match the branch condition. I'd either remove the mention of timestamps from the message, of specifically test if the type is a timestamp.